### DISCUSSION TITLE
solved

### DISCUSSION THREAD

**Discussion by LuckyzXz** [pyg-team/pytorch_geometric#9544] [Q&A]
I meet this problem when the model has been trained and I am applying it to test set.
The network structure of my model is as follows:

```python
class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels, metadata):
        super().__init__()
        self.conv1 = RGCNConv(in_channels=43, out_channels=hidden_channels, num_relations=len(metadata[1])) 
        self.conv2 = RGCNConv(in_channels=hidden_channels, out_channels=2, num_relations=len(metadata[1]))
        self.metadata = metadata

    def forward(self, x_dict, edge_index_dict, edge_type, batch):
        h = self.conv1(x_dict, edge_index_dict, edge_type)
        h = F.relu(h)
        h = self.conv2(h, edge_index_dict, edge_type)
        h = global_mean_pool(h,batch)
        return h

When the subgraph was input separately:

```python
with torch.no_grad():
    test1 = aligned_subgraphs[0].to_homogeneous()
    out = model(test1.x, test1.edge_index, test1.edge_type, test1.batch)
    print(out)
    # output: tensor([[-0.0821,  0.0705]])

with torch.no_grad():
    test2 = aligned_subgraphs[1].to_homogeneous()
    out = model(test2.x, test2.edge_index, test2.edge_type, test2.batch)
    print(out)
    # output: tensor([[0.0984, 0.2266]])
```

When they were inputed together in a batch:

```python
with torch.no_grad():
    loader = DataLoader(aligned_subgraphs[0:2], batch_size=2, shuffle=False)
    for batch in loader:
        batch = batch.to_homogeneous()
        out = model(batch.x, batch.edge_index, batch.edge_type, batch.batch)
        print(out)
        # tensor([[-0.0821,  0.0705],
                      [ 0.2211,  0.1975]])
```

The structure of aligned_subgraphs[0] and aligned_subgraphs[1] are here:

```
HeteroData(
  y=0,
  card_pid={ x=[1, 43] },
  token_pid={ x=[1, 43] },
  token_phone={ x=[1, 43] },
  token_email={ x=[1, 43] },
  pid_card={ x=[0, 43] },
  card_token={ x=[0, 43] },
  card_phone={ x=[0, 43] },
  card_email={ x=[0, 43] },
  token_card={ x=[0, 43] },
  (token_pid, to, token_phone)={ edge_index=[2, 1] },
  (token_pid, to, token_email)={ edge_index=[2, 1] },
  (card_pid, to, token_pid)={ edge_index=[2, 1] },
  (token_phone, rev_to, token_pid)={ edge_index=[2, 1] },
  (token_email, rev_to, token_pid)={ edge_index=[2, 1] },
  (token_pid, rev_to, card_pid)={ edge_index=[2, 1] },
  (card_pid, to, pid_card)={ edge_index=[2, 0] },
  (card_pid, to, card_token)={ edge_index=[2, 0] },
  (card_pid, to, card_phone)={ edge_index=[2, 0] },
  (card_pid, to, card_email)={ edge_index=[2, 0] },
  (pid_card, rev_to, card_pid)={ edge_index=[2, 0] },
  (card_token, rev_to, card_pid)={ edge_index=[2, 0] },
  (card_phone, rev_to, card_pid)={ edge_index=[2, 0] },
  (card_email, rev_to, card_pid)={ edge_index=[2, 0] },
  (token_pid, to, token_card)={ edge_index=[2, 0] },
  (token_card, rev_to, token_pid)={ edge_index=[2, 0] }
)

HeteroData(
  y=0,
  card_pid={ x=[1, 43] },
  pid_card={ x=[1, 43] },
  token_pid={ x=[1, 43] },
  token_phone={ x=[1, 43] },
  token_email={ x=[1, 43] },
  card_token={ x=[1, 43] },
  card_phone={ x=[1, 43] },
  card_email={ x=[1, 43] },
  token_card={ x=[0, 43] },
  (card_pid, to, pid_card)={ edge_index=[2, 1] },
  (token_pid, to, token_phone)={ edge_index=[2, 1] },
  (token_pid, to, token_email)={ edge_index=[2, 1] },
  (card_pid, to, token_pid)={ edge_index=[2, 1] },
  (card_pid, to, card_token)={ edge_index=[2, 1] },
  (card_pid, to, card_phone)={ edge_index=[2, 1] },
  (card_pid, to, card_email)={ edge_index=[2, 1] },
  (pid_card, rev_to, card_pid)={ edge_index=[2, 1] },
  (token_phone, rev_to, token_pid)={ edge_index=[2, 1] },
  (token_email, rev_to, token_pid)={ edge_index=[2, 1] },
  (token_pid, rev_to, card_pid)={ edge_index=[2, 1] },
  (card_token, rev_to, card_pid)={ edge_index=[2, 1] },
  (card_phone, rev_to, card_pid)={ edge_index=[2, 1] },
  (card_email, rev_to, card_pid)={ edge_index=[2, 1] },
  (token_pid, to, token_card)={ edge_index=[2, 0] },
  (token_card, rev_to, token_pid)={ edge_index=[2, 0] }
)
```

As you see, the output is different. But the model and input are same. The problem is still exist when i change the batch_size.
